errgrpc/grpc.go

```diff
@@ -47,47 +55,108 @@ func ToGRPC(err error) error {
 		return nil
 	}
 
-	if isGRPCError(err) {
+	if _, ok := status.FromError(err); ok {
 		// error has already been mapped to grpc
 		return err
 	}
 
+	s, extra := findStatus(err)
+	if s != nil {
+		var details []protoadapt.MessageV1
+
+		for _, e := range extra {
+			// Do not double encode proto messages, otherwise use Any
+			if pm, ok := e.(protoadapt.MessageV1); ok {
+				details = append(details, pm)
+			} else if pm, ok := e.(proto.Message); ok {
+				details = append(details, protoadapt.MessageV1Of(pm))
+			} else {
+
+				if reflect.TypeOf(e).Kind() == reflect.Ptr {
+					a, aerr := typeurl.MarshalAny(e)
+					if aerr == nil {
+						details = append(details, &anypb.Any{
+							TypeUrl: a.GetTypeUrl(),
+							Value:   a.GetValue(),
+						})
+						continue
+					}
+				}
+				if gs, ok := status.FromError(ToGRPC(e)); ok {
+					details = append(details, gs.Proto())
+				}
+				// TODO: Else include unknown extra error type
+			}
+		}
+
+		if len(details) > 0 {
+			if ds, _ := s.WithDetails(details...); ds != nil {
+				s = ds
+			}
+		}
+		err = s.Err()
+	}
+
+	return err
+}
+
+// findStatus finds the first error which matches a GRPC status and returns
+// any extra
+func findStatus(err error) (*status.Status, []error) {
+	multierr, ok := err.(interface{ Unwrap() []error })
```

Comment:
Do you need to traverse down here? Similar to https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/errors/wrap.go;l=17. I think there's a case where it gets wrapped, then we need traverse `Unwrap() error` to resolve `Unwrap() []error`.

Reply:
This is attempting to a depth first traversal to get the first  "original" error but we should probably check `Unwrap`. The `Is*` functions will traverse all the way down but once they are called, they won't pick up anymore `extra` errors. I can make a test case for this and maybe should try to make a graph diagram.

So we still should handle a case such as `errors.Join(fmt.Errorf("Some higher error: %w", errors.Join(err1, err2))` and in the current logic err2 will not be discarded.